File: frontend/src/features/reservations/types/reservation.ts
import { Space } from '../../spaces/types/space';
import { User } from '../../identity/types/user';
import { Subject } from './subject';

/**
 * Representa los diferentes estados transaccionales por los que puede atravesar el ciclo de vida de una reserva.
 */
export type ReservationStatus = 'SOLICITADA' | 'APROBADA' | 'RECHAZADA' | 'CANCELADA' | 'BLOQUEO';

/**
 * Representa las clasificaciones formales o tipologías válidas de las actividades asociadas a una reserva.
 */
export type ReservationType = 'CLASE' | 'EXAMEN' | 'OTRO';

/**
 * Estructura de datos integral que modela una reserva de espacios dentro del sistema.
 * 
 * @property id - Identificador único de la entidad reserva.
 * @property spaces - Colección de espacios físicos asociados e implicados en la reserva.
 * @property user - Entidad del usuario que figura como el solicitante o titular de la reserva.
 * @property startTime - Fecha y hora exactas correspondientes al comienzo de la actividad, en formato ISO 8601.
 * @property endTime - Fecha y hora exactas de la finalización de la actividad, en formato ISO 8601.
 * @property status - Estado de aprobación actual y vigente de la reserva.
 * @property type - Clasificación de la índole o naturaleza del evento a desarrollar.
 * @property title - Denominación descriptiva y opcional para la actividad objeto de la reserva.
 * @property subject - Asignatura académica opcionalmente vinculada al propósito de la reserva.
 * @property description - Bloque textual opcional con justificaciones o detalles técnicos adicionales.
 * @property responsible - Usuario del sistema asignado formalmente como responsable principal del evento.
 * @property responsibleName - Designación explícita de la persona encargada en los casos donde no se asigne a un usuario registrado en el sistema.
 * @property rejectionReason - Justificación textual proporcionada por los administradores en el caso de denegación de la solicitud.
 * @property createdAt - Fecha de creación del registro original en formato ISO 8601.
 */
export interface Reservation {
    id: number;
    spaces: Space[];
    user: User;
    startTime: string;
    endTime: string;
    status: ReservationStatus;
    type: ReservationType;
    title?: string;
    subject?: Subject;
    description?: string;
    responsible?: User;
    responsibleName?: string;
    rejectionReason?: string;
    createdAt: string;
}

File: backend/src/main/java/com/tfg/backend/modules/reservation/model/Reservation.java
package com.tfg.backend.modules.reservation.model;
import com.tfg.backend.modules.space.model.Space;
import com.tfg.backend.modules.identity.model.User;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;
import java.util.Set;
import java.util.HashSet;

/**
 * Entidad principal del dominio que representa una reserva formalizada sobre los espacios de la institución.
 * Establece la vinculación entre una franja cronológica específica, un conjunto de recursos espaciales y,
 * de manera opcional, el marco curricular de una asignatura. Soporta un ciclo de vida de estados y 
 * contempla la figura de bloqueos técnicos para tareas de mantenimiento.
 */
@Entity
@Table(name = "reservations")
@Getter
@Setter
@ToString(exclude = {"spaces", "user", "subject"})
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Reservation {

    /** Identificador único de la reserva. */
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    /**
     * El conjunto de espacios físicos reservados simultáneamente en la misma solicitud.
     * Su carga temprana (EAGER) resulta indispensable para facilitar su rápida proyección gráfica
     * en el módulo de calendario o cuadrícula horaria.
     */
    @ManyToMany(fetch = FetchType.EAGER)
    @JoinTable(
        name = "reservation_spaces",
        joinColumns = @JoinColumn(name = "reservation_id"),
        inverseJoinColumns = @JoinColumn(name = "space_id")
    )
    @Builder.Default
    private Set<Space> spaces = new HashSet<>();

    /** 
     * Usuario propietario y emisor originario de la solicitud de reserva. 
     */
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "user_id", nullable = false)
    private User user;

    /** 
     * Usuario registrado que asume la responsabilidad funcional de la actividad, en caso de delegación. 
     */
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "responsible_user_id")
    private User responsible;

    /** 
     * El momento cronológico en el que da comienzo el uso de los espacios. 
     */
    @Column(nullable = false)
    private LocalDateTime startTime;

    /** 
     * El momento cronológico que dictamina el término de la reserva estipulada. 
     */
    @Column(nullable = false)
    private LocalDateTime endTime;

    /** 
     * Situación administrativa actual en la que se encuentra el ciclo de vida de la reserva. 
     */
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ReservationStatus status;

    /** 
     * Tipología académica o administrativa adjudicada a la ocupación propuesta. 
     */
    @Enumerated(EnumType.STRING)
    @Builder.Default
    private ReservationType type = ReservationType.OTRO;

    /** 
     * Enunciado o título representativo que designa la naturaleza o propósito del evento. 
     */
    private String title;

    /** 
     * Referencia a la entidad de la asignatura en caso de reservas curriculares (requisito para docencia o evaluación). 
     */
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "subject_id")
    private Subject subject;

    /** 
     * Aclaraciones funcionales complementarias o exigencias logísticas especiales. 
     */
    @Column(length = 1000)
    private String description;

    /** 
     * Designación nominativa del docente o personal a cargo de la gestión in situ. 
     */
    private String responsibleName;

    /** 
     * Argumentación formal provista por los órganos gestores al denegar la solicitud. 
     */
    private String rejectionReason;

    /** 
     * Indicador lógico de control que dictamina si se ha procedido al envío automatizado de notificaciones previas. 
     */
    @Builder.Default
    private boolean reminderSent = false;

    /** 
     * Indicador de supervisión para el sistema subyacente responsable de enviar avisos de solicitudes demoradas. 
     */
    @Builder.Default
    private Boolean approvalReminderSent = false;

    /**
     * Comprueba si se ha enviado el recordatorio de aprobación.
     *
     * @return true si se envió, false en caso contrario
     */
    public boolean isApprovalReminderSent() {
        return approvalReminderSent != null && approvalReminderSent;
    }

    /** 
     * Marca temporal que documenta fehacientemente el instante de registro inicial del objeto en el sistema. 
     */
    private LocalDateTime createdAt;

    /**
     * Establece la fecha de creación antes de persistir.
     */
    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
    }
}
